### PR TITLE
[JSC] Wasm repatch should be maintained by reversed data structure

### PR DESCRIPTION
**Pull request by Constellation**
#### 7ab897711f80e7c2a341ec065a1ecda29bdeac3b
<pre>
[JSC] Wasm repatch should be maintained by reversed data structure
<a href="https://bugs.webkit.org/show_bug.cgi?id=251316">https://bugs.webkit.org/show_bug.cgi?id=251316</a>
rdar://104775515

Reviewed by Saam Barati.

When repatching wasm callsites with a newly compiled code, we are literally iterating every callsites on the wasm module,
and updating them with a new one if it is matching against the registered one. But this is too stupid and it becomes exponentially
slow when wasm module gets large.

We should just maintain the data structure which answers callsites calling a function X. And when X is recompiled (e.g. -&gt; BBQ -&gt; OMG),
then we should query to this data structure to obtain the list of callsites calling to X super quickly.

This patch adds Wasm::CallsiteCollection. This maintains Vector&lt;Vector&gt; basically, Vector of callsites calling a function X.

This improves large wasm website performance in a dramatic way.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::work):
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::CalleeGroup):
* Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
* Source/JavaScriptCore/wasm/WasmCallsiteCollection.cpp: Added.
(JSC::Wasm::CallsiteCollection::addCallsites):
(JSC::Wasm::CallsiteCollection::addCalleeGroupCallsites):
(JSC::Wasm::CallsiteCollection::updateCallsitesToCallUs):
* Source/JavaScriptCore/wasm/WasmCallsiteCollection.h: Added.
(JSC::Wasm::CallsiteCollection::CallsiteCollection):
(JSC::Wasm::CallsiteCollection::calleeGroupCallsites const):
* Source/JavaScriptCore/wasm/WasmCompilationMode.h:
* Source/JavaScriptCore/wasm/WasmOMGPlan.cpp:
(JSC::Wasm::OMGPlan::work):
* Source/JavaScriptCore/wasm/WasmOSREntryPlan.cpp:
(JSC::Wasm::OSREntryPlan::work):
* Source/JavaScriptCore/wasm/WasmPlan.cpp:
(JSC::Wasm::Plan::updateCallSitesToCallUs): Deleted.
* Source/JavaScriptCore/wasm/WasmPlan.h:

Canonical link: <a href="https://commits.webkit.org/259529@main">https://commits.webkit.org/259529@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/043b5c7105a6873d268adad54c8535ac30d7775d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105170 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14253 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/38055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114429 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174607 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15397 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/5172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97484 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110926 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/38055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/38055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/94992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7584 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/38055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/93052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5314 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7679 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/5172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/93052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13731 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/38055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101754 "Built successfully") | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9467 "Failed to checkout and rebase branch from PR 9281") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/25387 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3510 "Built successfully and passed tests") | | | |

### CODE REVIEW

File: Source/JavaScriptCore/wasm/WasmRepatchTree.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+#pragma once
+
+#if ENABLE(WEBASSEMBLY)
+
+namespace JSC { namespace Wasm {
+
+class CalleeGroup;
+
+class RepatchTree {
```
Comment (saambarati) [inline Source/JavaScriptCore/wasm/WasmRepatchTree.h]: But: This isn’t really a tree. Maybe we can pick a different name

**Review by saambarati — APPROVED** [review]
r=me

File: Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp [hunk 2]
```diff
@@ -52,16 +52,20 @@ CalleeGroup::CalleeGroup(MemoryMode mode, const CalleeGroup& other)
     , m_llintCallees(other.m_llintCallees)
     , m_embedderCallees(other.m_embedderCallees)
     , m_wasmIndirectCallEntryPoints(other.m_wasmIndirectCallEntryPoints)
-    , m_wasmToWasmCallsites(other.m_wasmToWasmCallsites)
     , m_wasmToWasmExitStubs(other.m_wasmToWasmExitStubs)
+    , m_repatchTree(m_calleeCount)
 {
+    Locker locker { m_lock };
+    auto callsites = other.repatchTree().calleeGroupCallsites();
```
Comment (saambarati) [inline Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp]: Do we need a temporary variable here?

File: Source/JavaScriptCore/wasm/WasmRepatchTree.h (hunk 1 above)
Comment (Constellation) [reply Source/JavaScriptCore/wasm/WasmRepatchTree.h]: I've renamed it to just CallsitesCollection

File: Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp (hunk 2 above)
Comment (Constellation) [reply Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp]: Unfortunately, yes, to make WTFMove work.

**Comment by Constellation** [comment]
Just renaming. And now JSC gets compiled (jsc-arm64). Landing.

**Pull request merged**